### PR TITLE
Implement variational free energy function (#297)

### LINKED ISSUE #297: Add variational free energy computation
Add a method that computes the variational free energy for a given pair of (generative model, variational posterior).

A few different options that I can imagine

- computing the 'single timestep' variational free energy (effectively computed currently inside the vanilla fixed-point iteration algorithm) using only the current posterior over hidden states, past action, B matrix, A matrix and priors over A and B. There is a version of this already in the `numpy` backend ([pymdp.legacy.maths::calc_free_energy()](https://github.com/infer-actively/pymdp/blob/fcd483fc5dd8c896af4ca1cebdd3e8c9e0c80800/pymdp/legacy/maths.py#L396C5-L396C21)) but it needs to be generalized and KL divergences for parameter posteriors/priors need to be added
- computing the full variational free energy given a sequence of observations, actions, hidden state posteriors (indexed from t=0 to t=current timestep) and parameter posteriors/priors.

In the case of the latter (full free energy), I guess we should have some check that makes sure that if the inference algorithm is "ovf" or "fpi", that smoothing should be performed before computing the full variational free energy.

### PR DESCRIPTION
**Pull request by conorheins**
## Summary
- add `pymdp.maths.calc_vfe(...)` for canonical single-step and sequence variational free energy, including optional Dirichlet KL terms and a `joint_qs` path for smoothed sequence VFE
- plumb `return_info=True` through `pymdp.inference.update_posterior_states(...)` and `Agent.infer_states(...)` so callers can opt into `vfe_t`, `vfe`, and component diagnostics without changing default return values
- refactor sequence VFE computation onto `lax.scan`, add direct JIT coverage for `calc_vfe(...)` with and without `joint_qs`, and remove the obsolete `compute_free_energy(...)` wrapper
- centralize default dependency resolution in `pymdp.utils`, reuse it from math/inference/control/agent paths, and tighten action-history validation so multi-factor histories must be explicit `(T-1, num_factors)` arrays
- preserve sequence transition information when `past_actions=None` but `B` has singleton control dimensions, while raising clear errors when multi-action sequence inference or VFE diagnostics are requested without action histories
- update the neural encoder notebook to consume the canonical VFE diagnostics API instead of its local one-step helper, and verify it still reproduces the same key reported metrics
- add regression coverage for legacy one-step agreement, smoothed exact/OVF sequence VFE, gradient finiteness with and without `joint_qs`, action-history validation, singleton-control missing-action paths, missing-action sequence windows, batched agent diagnostics, and monotone FPI VFE descent

Closes #297
Related to #195

## Testing
- `python3 -m py_compile pymdp/utils.py pymdp/maths.py pymdp/inference.py pymdp/agent.py pymdp/control.py test/test_vfe_jax.py`
- `python -m pytest -o addopts='' test/test_vfe_jax.py`
- `python -m pytest -o addopts='' test/test_hmm_associative_scan.py -k "exact_inference_accepts_missing_past_actions or exact_inference_matches_wrapper"`
- `python -m pytest -o addopts='' test/test_agent_jax.py -k "smoothing_ovf_updates_A_when_learn_B_false"`
- re-executed `examples/advanced/pymdp_with_neural_encoder.ipynb` and confirmed the saved key metrics still match (aside from wall-clock runtime)

### CODE REVIEW

File: pymdp/inference.py [hunk 1]
```diff
@@ -173,7 +174,8 @@ def update_posterior_states(
     distr_obs: bool = True,
     inference_horizon: int | None = None,
     valid_steps: int | Array | None = None,
-) -> list[Array]:
+    return_info: bool = False,
+) -> list[Array] | tuple[list[Array], dict[str, Array]]:
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/inference.py]: using a `TypedDict` now for that one

```python
class VFEInfo(TypedDict):
    vfe_t: ArrayLike
    vfe: ArrayLike
    vfe_components: dict[str, ArrayLike]
```

File: pymdp/maths.py [hunk 2]
```diff
@@ -297,25 +394,33 @@ def compute_accuracy(qs: list[ArrayLike], obs: list[ArrayLike], A: list[ArrayLik
...
     ArrayLike
         Expected log-likelihood term.
     """
+    A_dependencies = _default_a_dependencies(len(qs), len(A), A_dependencies)
+    log_likelihoods = compute_log_likelihood_per_modality(obs, A, distr_obs=distr_obs)
+    accuracy = 0.0
 
-    log_likelihood = compute_log_likelihood(obs, A)
+    for ll_m, deps_m in zip(log_likelihoods, A_dependencies):
+        accuracy += _expected_log_prob(ll_m, [qs[f] for f in deps_m])
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py]: no that seems too long -- that will get tripped earlier in agent init anyway

File: pymdp/maths.py [hunk 3]
```diff
@@ -335,22 +440,354 @@ def compute_free_energy(
...
+    if obs is not None and A is None:
+        raise ValueError("`A` must be provided when `obs` is provided")
+
+    A_dependencies = (
+        None if A is None else _default_a_dependencies(num_factors, len(A), A_dependencies)
+    )
+    B_dependencies = (
+        None if B is None else _default_b_dependencies(num_factors, B_dependencies)
+    )
+
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py:539]: nope

File: pymdp/inference.py [hunk 4]
```diff
@@ -159,6 +166,162 @@ def _select_transitions_for_actions(b: ArrayLike, action_idx: Array) -> Array:
...
+        "obs": obs,
+        "A": A,
+        "B": B_for_vfe if past_actions is not None else None,
+        "past_actions": past_actions,
+        "A_dependencies": A_dependencies,
+        "B_dependencies": B_dependencies,
+        "obs_valid_mask": obs_valid_mask,
+        "transition_valid_mask": transition_valid_mask,
+        "distr_obs": distr_obs,
+    }
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/inference.py:336]: now should be addressed by 057d0a2

File: pymdp/inference.py [hunk 5]
```diff
@@ -159,6 +166,162 @@ def _select_transitions_for_actions(b: ArrayLike, action_idx: Array) -> Array:
...
+            )
+
+    if valid_steps is not None:
+        obs_valid_mask, transition_valid_mask = _build_sequence_validity_masks(
+            obs, past_actions, valid_steps
+        )
+
+    conditioned_B = None
+    if past_actions is not None:
+        conditioned_B = _condition_transitions_on_actions(B, past_actions)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/inference.py:230]: addressed by 057d0a2

File: pymdp/maths.py [hunk 6]
```diff
@@ -335,22 +441,347 @@ def compute_free_energy(
...
+    if not is_sequence:
+        neg_entropy_t = _neg_entropy_term(qs)
+        prior_cross_entropy_t = _state_prior_term(qs)
+        accuracy_t = _accuracy_term(qs, obs)
+        transition_cross_entropy_t = jnp.array(0.0)
+        vfe_t = neg_entropy_t + prior_cross_entropy_t - accuracy_t
+    else:
+        T = qs[0].shape[0]
+        past_actions = _ensure_vfe_action_history_shape(past_actions, num_factors)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py:579]: addressed this by enforcing the `(T, num_factors)`-shaped `past_actinos`

File: pymdp/maths.py [hunk 7]
```diff
@@ -297,60 +383,378 @@ def compute_accuracy(qs: list[ArrayLike], obs: list[ArrayLike], A: list[ArrayLik
...
+        raise ValueError("`A` must be provided when `obs` is provided")
 
-    return vfe
+    A_dependencies = (
+        None if A is None else resolve_a_dependencies(num_factors, len(A), A_dependencies)
+    )
+    B_dependencies = (
+        None if B is None else resolve_b_dependencies(num_factors, B_dependencies)
+    )
+
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py:539]: irrelevant / not that unsafe as you suggest

File: pymdp/maths.py [hunk 8]
```diff
@@ -297,60 +383,378 @@ def compute_accuracy(qs: list[ArrayLike], obs: list[ArrayLike], A: list[ArrayLik
...
     ArrayLike
         Expected log-likelihood term.
     """
+    A_dependencies = resolve_a_dependencies(len(qs), len(A), A_dependencies)
+    log_likelihoods = compute_log_likelihood_per_modality(obs, A, distr_obs=distr_obs)
+    accuracy = 0.0
 
-    log_likelihood = compute_log_likelihood(obs, A)
+    for ll_m, deps_m in zip(log_likelihoods, A_dependencies):
+        accuracy += _expected_log_prob(ll_m, [qs[f] for f in deps_m])
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py:404]: not that unsafe since most of the time you'll be using this with agent.infer_states() where that validation has already happened in the __init__ of agent

File: pymdp/utils.py [hunk 9]
```diff
@@ -178,15 +207,14 @@ def random_B_array(
...
-    if B_dependencies is None:
-        B_dependencies = [[f] for f in range(num_factors)]
+    B_dependencies = resolve_b_dependencies(num_factors, B_dependencies)
 
     if B_action_dependencies is None:
         assert len(num_controls) == len(num_states)
-        B_action_dependencies = [[f] for f in range(num_factors)]
     else:
         unique_controls = list(set(sum(B_action_dependencies, [])))        
         assert unique_controls == list(range(len(num_controls)))
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/utils.py]: fair point, now addressed with f38cc11

File: pymdp/maths.py [hunk 10]
```diff
@@ -297,60 +383,381 @@ def compute_accuracy(qs: list[ArrayLike], obs: list[ArrayLike], A: list[ArrayLik
...
+                    start_scale * neg_entropy_marginal_t
+                    + (1.0 - start_scale) * neg_entropy_cond_t
+                )
+
+            if B is None:
+                transition_term_t = jnp.array(0.0, dtype=neg_entropy_term_t.dtype)
+            else:
+                transition_term_t = 0.0
+                actions_t = None if padded_past_actions is None else padded_past_actions[t]
+                for f in range(max(len(B), len(B_dependencies))):
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py:684]: I dont want to clog up the code with more of those assertions, so i think we should just eat the cost of the 'confusing' indexerror

File: pymdp/maths.py [hunk 11]
```diff
@@ -297,60 +383,381 @@ def compute_accuracy(qs: list[ArrayLike], obs: list[ArrayLike], A: list[ArrayLik
...
     """
+    A_dependencies = resolve_a_dependencies(len(qs), len(A), A_dependencies)
+    log_likelihoods = compute_log_likelihood_per_modality(obs, A, distr_obs=distr_obs)
+    accuracy = 0.0
 
-    log_likelihood = compute_log_likelihood(obs, A)
+    for m in range(max(len(log_likelihoods), len(A_dependencies))):
+        ll_m = log_likelihoods[m]
+        deps_m = A_dependencies[m]
+        accuracy += _expected_log_prob(ll_m, [qs[f] for f in deps_m])
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/maths.py:403]: don't mind, let's eat the cost of the confusing `IndexError`

File: test/test_utils_jax.py [hunk 12]
```diff
@@ -127,7 +127,27 @@ def test_random_B_array_defaults_to_self_dependencies(self):
...
-    
+
+    def test_random_B_array_accepts_out_of_order_control_dependency_lists(self):
+        """`random_B_array` should validate control coverage independent of list ordering"""
+
+        key = jr.PRNGKey(17)
+        num_states = [3, 2]
+        num_controls = [2, 3]
+        B_dependencies = [[0], [1]]
+        B_action_dependencies = [[1], [1, 0]]
```
[the remark this answers is not in the document]
Comment (conorheins) [reply test/test_utils_jax.py:138]: Good point, now addressed with corrected ordering assumptions in [beb8563](https://github.com/infer-actively/pymdp/pull/370/commits/beb85633b9d0f55fa8af04ffae5a86290a9ccd46), now we construct lagging shapes based on given dep order of B_action_deps or B-deps

File: pymdp/utils.py [hunk 13]
```diff
@@ -178,15 +207,14 @@ def random_B_array(
...
 
     if B_action_dependencies is None:
         assert len(num_controls) == len(num_states)
-        B_action_dependencies = [[f] for f in range(num_factors)]
     else:
-        unique_controls = list(set(sum(B_action_dependencies, [])))        
-        assert unique_controls == list(range(len(num_controls)))
+        unique_controls = set(sum(B_action_dependencies, []))
+        assert unique_controls == set(range(len(num_controls)))
+    B_action_dependencies = resolve_b_action_dependencies(num_factors, B_action_dependencies)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/utils.py:217]: addressed by [beb8563](https://github.com/infer-actively/pymdp/pull/370/commits/beb85633b9d0f55fa8af04ffae5a86290a9ccd46), now we construct shapes in the order given by the dep lists
